File: daily_problems/2025/06/0620/personal_submission/cf183c_yawn_sean.py
# Submission link: https://codeforces.com/contest/183/submission/325213280
def main():
    n, m = MII()
    path = [[] for _ in range(n)]

    for _ in range(m):
        u, v = GMI()
        path[u].append(v)
        path[v].append(~u)

    note = 10 ** 6
    cols = [note] * n

    ans = 0

    for i in range(n):
        if cols[i] == note:
            cols[i] = 0
            
            stk = [i]
            while stk:
                u = stk.pop()
                
                for v in path[u]:
                    if v >= 0:
                        if cols[v] == note:
                            cols[v] = cols[u] + 1
                            stk.append(v)
                        else:
                            ans = math.gcd(ans, cols[u] + 1 - cols[v])
                    else:
                        v = ~v
                        if cols[v] == note:
                            cols[v] = cols[u] - 1
                            stk.append(v)
                        else:
                            ans = math.gcd(ans, cols[u] - 1 - cols[v])

    ans = abs(ans)

    print(ans if ans else n)

File: daily_problems/2025/06/0620/personal_submission/cf183c_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    vector<vector<pair<int, int>>> path(n);
    while (m --) {
        int u, v;
        cin >> u >> v;
        u --, v --;
        path[u].emplace_back(v, 1);
        path[v].emplace_back(u, -1);
    }

    int note = 1e6, ans = 0;
    vector<int> cols(n, note);

    auto dfs = [&] (auto &self, int u) -> void {
        for (auto &[v, w]: path[u]) {
            if (cols[v] == note) {
                cols[v] = cols[u] + w;
                self(self, v);
            }
            else ans = gcd(ans, abs(cols[u] + w - cols[v]));
        }
    };

    for (int i = 0; i < n; i ++) {
        if (cols[i] == note) {
            cols[i] = 0;
            dfs(dfs, i);
        }
    }

    cout << (ans ? ans : n);

    return 0;
}